xarray/ufuncs.py

```diff
@@ -0,0 +1,125 @@
+"""xarray specific universal functions."""
+
+import textwrap
+import warnings
+
+import numpy as np
+
+import xarray as xr
+from xarray.core.groupby import GroupBy
+from xarray.namedarray.pycompat import array_type
+
+
+def _walk_array_namespaces(obj, namespaces):
+    if isinstance(obj, xr.DataTree):
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
```

Comment:
This is the right thing to do, but it may be worth noting that `apply_ufunc` does not work for DataTree yet: https://github.com/pydata/xarray/issues/9789

Reply:
Yep I noticed that, will make a note